### PR TITLE
KAFKA-5829: Only delete producer snapshots before the recovery point

### PR DESCRIPTION
**Pull request by ijuma**

### CODE REVIEW

**Comment by ijuma** [comment]
I haven't written tests yet, but I wanted to quickly prototype the approach suggested by @hachikuji in #3782: https://github.com/apache/kafka/pull/4023/commits/291c73e54b808d0c0af57c08107c3c8f357e8188 

Thoughts @lindong28, @hachikuji and @junrao?

**Comment by lindong28** [comment]
@ijuma Thanks for the patch. I think this patch addresses the problem and I am OK if we use this approach for 1.0 release. Let me explain why I think the other approach is better.

The approach in this patch is to only delete snapshot after the recovery offset. The other approach is to allow snapshot after recovery offset to be deleted, but do not rebuild snapshot whose offset is smaller than the latest existing snapshot during recovery. Both approach helps prevent broker from having to reload all segments after a hard kill in most cases. 

There is slight difference between the two. I personally prefer the approach in the other pull request for the following two reasons:

- By definition of the snapshot, only the snapshot with latest offset matters. Thus there is no reason to re-build snapshot when there exists snapshot with larger offset, unless the segment is corrupted. Thus this is something we should do even if we only delete snapshot after the recovery offset.

- The logic around the snapshots is a bit more complicated in this patch. Previously the latest two snapshots will be kept. Now we have to know the recovery offset when thinking about the remaining snapshot files. It is not a big deal. But it is better if we don't have to thinking about recovery offset when determining the snapshot.

**Comment by ijuma** [comment]
@lindong28 during recovery, you don't know which segments are corrupted until you read them. As explained by @hachikuji, that's the main issue with not rebuilding the producer snapshots as you recover a segment. What if you find a corrupted segment later? You'd have to do a second pass. I've mentioned that in a comment in this PR.

**Comment by lindong28** [comment]
@ijuma If a segment is corrupted, we would need to delete all snapshots after this segment and rebuild snapshot for this segment. But since we only delete snapshot after data is flushed to disk and we keep the last two snapshots, the second -to-last snapshot should be valid, i.e. all segments before this snapshot should not be corrupted. Thus no segment should be double scanned, right? Did I miss something here?

**Comment by ijuma** [comment]
@lindong28 If you have multiple segments and no snapshot, you need to scan all of them to build a snapshot. If you have an existing snapshot, you can incrementally update it. That is my understanding from @hachikuji's explanation. @hachikuji, please correct me if I'm wrong.

**Comment by hachikuji** [comment]
@ijuma Yes, that's correct. If we need to recover from an offset which is lower than the existing snapshots, then the only option is to scan from the beginning of the log.

File: core/src/main/scala/kafka/log/Log.scala [hunk 1]
```diff
@@ -1638,7 +1643,7 @@ object Log {
   /** a time index file */
   val TimeIndexFileSuffix = ".timeindex"
 
-  val PidSnapshotFileSuffix = ".snapshot"
+  val ProducerIdSnapshotFileSuffix = ".snapshot"
```
Comment (hachikuji) [inline core/src/main/scala/kafka/log/Log.scala]: Maybe we can drop `Id` from this name?

File: core/src/main/scala/kafka/log/LogManager.scala [hunk 2]
```diff
@@ -87,9 +87,14 @@ class LogManager(logDirs: Array[File],
   @volatile private var logStartOffsetCheckpoints = liveLogDirs.map(dir =>
     (dir, new OffsetCheckpointFile(new File(dir, LogStartOffsetCheckpointFile), logDirFailureChannel))).toMap
 
-  private def offlineLogDirs = logDirs.filterNot(_liveLogDirs.contains)
   private val preferredLogDirs = new ConcurrentHashMap[TopicPartition, String]()
 
+  private def offlineLogDirs: Iterable[File] = {
```
Comment (hachikuji) [inline core/src/main/scala/kafka/log/LogManager.scala:93]: Can you explain why this change was needed?

**Comment by lindong28** [comment]
@ijuma If you have no snapshot at all, with both approaches, you need to scan from the first segment to rebuild snapshot. There is no difference.

The problem here is that, if we want to recover from segment 100, and there exits snapshot with offset 300, do we even need to build snapshot with offset 100. I think the answer is no. Jason and I have a very long discussion around this in the other pull request. Since Jason insists the other approach does not work even after I answered all his questions, let’s just use this approach so that we address the issuse in 1.0 release.

**Review by hachikuji — COMMENTED** [review]
Thanks for the patch. I tested it locally and it seems to work as intended (i.e. it actually prevents loading from the beginning of the log).

One thing I happened to notice is that the recovery point does not get updated after completing log recovery. That means that recovery point will stay at whatever value was stored in the checkpoint file until we begin appending. I don't think it's a big deal, but maybe we may as well improve that here?

**Comment by ijuma** [comment]
@lindong28, there is a difference. Instead of this back and forth that is not fruitful, I will try to write a test.

File: core/src/main/scala/kafka/log/LogManager.scala (hunk 2 above)
Comment (ijuma) [reply core/src/main/scala/kafka/log/LogManager.scala:93]: `contains` on a queue is O(n), so the previous code was O(n^2). I thought that was a bit surprising given the method name so I replaced it with a O(n) implementation. Probably won't make much of a difference unless people have a lot of log directories, but this could happen.

**Comment by hachikuji** [comment]
LGTM. I think the failures in `ResetIntegrationTest` are unrelated. I will run them a couple times locally to be sure.
